tidy3d/plugins/mode/mode_solver.py

```diff
@@ -339,6 +353,339 @@ def data_raw(self) -> ModeSolverData:
+                for field_name in ("Ex", "Ey", "Ez", "Hx", "Hy", "Hz")
+            }
+
+            for field_type in ["E", "H"]:
+                field_data_cylindrical[field_type + "r"][i, 0] = (
+                    fields_at_point[field_type + cmp_1] * cos_theta
+                    + fields_at_point[field_type + cmp_2] * sin_theta
+                )
+                field_data_cylindrical[field_type + "theta"][i, 0] = (
+                    -fields_at_point[field_type + cmp_1] * sin_theta
+                    + fields_at_point[field_type + cmp_2] * cos_theta
+                )
+                field_data_cylindrical[field_type + "axial"][i, 0] = fields_at_point[
+                    field_type + axial_name
+                ]
+
+        coords = {
+            "rho": rho,
+            "theta": theta,
+            "axial": axial,
+            "f": f,
+            "mode_index": mode_index,
+        }
+
+        solver_data_cylindrical = {
+            name: ScalarModeFieldCylindricalDataArray(field_data_cylindrical[name], coords=coords)
+            for name in ("Er", "Etheta", "Eaxial", "Hr", "Htheta", "Haxial")
+        }
+        solver_data_cylindrical["n_complex"] = mode_solver_data.n_complex
+
+        return solver_data_cylindrical
+
+    def _mode_rotation(
+        self,
+        solver_ref_data_cylindrical: Dict[
+            Union[ScalarModeFieldCylindricalDataArray, ModeIndexDataArray]
+        ],
+        solver_temp_data: ModeSolverData,
+    ) -> ModeSolverData:
+        """rotate the mode solver solution from the reference plane in cylindrical coordinates
+        to the practical plane"""
+
+        # Extract coordinate values from one of the six field components
+        f, mode_index = solver_temp_data.Ex.f, solver_temp_data.Ex.mode_index
+        x, y, z = solver_temp_data.Ex.x, solver_temp_data.Ex.y, solver_temp_data.Ex.z
+        axial = solver_ref_data_cylindrical["Er"].axial
+
+        # Initialize output arrays
+        shape = (x.size, y.size, z.size, f.size, mode_index.size)
+        rotated_field_data = {
+            name: np.zeros(shape, dtype=complex) for name in ("Ex", "Ey", "Ez", "Hx", "Hy", "Hz")
+        }
+
+        # offset coors by bend_center
+        _, idx_uv = self.plane.pop_axis((0, 1, 2), axis=self.bend_axis_3d)
+        idx_u, idx_v = idx_uv
+
+        pts = [solver_temp_data.Ex[name].values.copy() for name in ("x", "y", "z")]
+        pts[idx_u] -= self.bend_center[idx_u]
+        pts[idx_v] -= self.bend_center[idx_v]
+
+        rho = np.sqrt(pts[idx_u] ** 2 + pts[idx_v] ** 2)
+        theta = np.arctan2(pts[idx_v], pts[idx_u])
+        theta_rel = theta - self.theta_reference
+
+        cos_theta = pts[idx_u] / rho
+        sin_theta = pts[idx_v] / rho
+
+        cmp_normal, source_names = self.plane.pop_axis(("x", "y", "z"), axis=self.bend_axis_3d)
+        cmp_1, cmp_2 = source_names
+
+        for mode_idx in range(mode_index.size):
+            for f_idx, f_val in enumerate(f.values):
```

Comment:
I would think it should be possible to vectorize if not along both of these dims, then at least over the frequency one. This could be left as an optimization for later though.